Sources/Networking/v2/APIRequestV2.swift

```diff
@@ -0,0 +1,78 @@
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+public struct APIRequestV2: CustomDebugStringConvertible {
+
+    public typealias QueryItems = [String: String]
+
+    let timeoutInterval: TimeInterval
+    let responseConstraints: [APIResponseConstraints]?
+    public let urlRequest: URLRequest
+
+    /// Designated initialiser
+    /// - Parameters:
+    ///   - url: The request URL, included protocol and host
+    ///   - method: HTTP method
+    ///   - queryItems: A key value dictionary with query parameters
+    ///   - headers: HTTP headers
+    ///   - body: The request body
+    ///   - timeoutInterval: The request timeout interval, default is `60`s
+    ///   - cachePolicy: The request cache policy, default is `.useProtocolCachePolicy`
+    ///   - responseRequirements: The response requirements
+    ///   - allowedQueryReservedCharacters: The characters in this character set will not be URL encoded in the query parameters
+    public init?(url: URL,
+                 method: HTTPRequestMethod = .get,
+                 queryItems: QueryItems? = nil,
+                 headers: APIRequestV2.HeadersV2? = APIRequestV2.HeadersV2(),
+                 body: Data? = nil,
+                 timeoutInterval: TimeInterval = 60.0,
+                 cachePolicy: URLRequest.CachePolicy? = nil,
+                 responseConstraints: [APIResponseConstraints]? = nil,
+                 allowedQueryReservedCharacters: CharacterSet? = nil) {
+        self.timeoutInterval = timeoutInterval
+        self.responseConstraints = responseConstraints
+
+        // Generate URL request
+        guard var urlComps = URLComponents(url: url, resolvingAgainstBaseURL: true) else {
+            return nil
+        }
+        urlComps.queryItems = queryItems?.toURLQueryItems(allowedReservedCharacters: allowedQueryReservedCharacters)
+        guard let finalURL = urlComps.url else {
+            return nil
+        }
+        var request = URLRequest(url: finalURL, timeoutInterval: timeoutInterval)
+        request.allHTTPHeaderFields = headers?.httpHeaders
+        request.httpMethod = method.rawValue
+        request.httpBody = body
+        if let cachePolicy = cachePolicy {
+            request.cachePolicy = cachePolicy
+        }
+        self.urlRequest = request
+    }
+
+    public var debugDescription: String {
+        """
+        \(urlRequest.httpMethod ?? "Nil") \(urlRequest.url?.absoluteString ?? "nil")
+        Headers: \(urlRequest.allHTTPHeaderFields?.debugDescription ?? "-")
+        Body: \(urlRequest.httpBody?.debugDescription ?? "-")
+        Constraints: \(responseConstraints?.debugDescription ?? "-")
+        """
+    }
```

Comment:
Nice idea to add this. While we are here, maybe we can add all properties, i.e:

```suggestion
   public var debugDescription: String {
        """
        APIRequestV2:
        URL: \(urlRequest.url?.absoluteString ?? "nil")
        Method: \(urlRequest.httpMethod ?? "Nil")
        Headers: \(urlRequest.allHTTPHeaderFields?.debugDescription ?? "-")
        Body: \(urlRequest.httpBody?.debugDescription ?? "-")
        Timeout Interval: \(timeoutInterval)s
        Cache Policy: \(urlRequest.cachePolicy)
        Response Constraints: \(responseConstraints?.map { $0.rawValue } ?? [])
        """
    }
```

Reply:
Sure! thanks